plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -666,17 +712,145 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
+  return sp
+getSourceSpan _ _ = Nothing
+
+-- | Obviously this function computes a GHC.RealSrcSpan from a CovLoc
+toCovLoc :: GHC.RealSrcSpan -> CovLoc
+toCovLoc sp = CovLoc (GHC.unpackFS $ GHC.srcSpanFile sp)
+                     (GHC.srcSpanStartLine sp)
+                     (GHC.srcSpanEndLine sp)
+                     (GHC.srcSpanStartCol sp)
+                     (GHC.srcSpanEndCol sp)
+
+-- Here be dragons:
+-- See Note [Tracking coverage and lazyness]
+-- See Note [Coverage order]
+-- | Annotate a term for coverage
+coverageCompile :: CompilingDefault uni fun m
+                => GHC.CoreExpr -- ^ The original expression
+                -> GHC.Type -- ^ The type of the expression
+                -> GHC.RealSrcSpan -- ^ The source location of this expression
+                -> PIRTerm uni fun -- ^ The current term (this is what we add coverage tracking to)
+                -> CoverageType -- ^ The type of coverage to do next
+                -> m (PIRTerm uni fun)
+coverageCompile originalExpr exprType src compiledTerm covT =
+  case covT of
+    -- Add a location covereage annotation to tell us "we've executed this piece of code"
+    LocationCoverage -> do
+      ann <- addLocationToCoverageIndex (toCovLoc src)
+      ty <- compileTypeNorm exprType
+      mkLazyTrace ty (T.pack . show $ ann) compiledTerm
+
+    -- Add two boolean coverage annotations to tell us "this boolean has been True/False respectively"
+    BooleanCoverage -> do
+      -- Check if the thing we are compiling is a boolean
+      CompileContext {ccNameInfo=nameInfo} <- ask
+      let bool = GHC.getName <$> Map.lookup ''Bool nameInfo -- TODO: This should always return a `Just` and it may be worthwhile to crash the compiler with an error if it doesn't
+          tyHeadName = GHC.getName <$> GHC.tyConAppTyCon_maybe exprType
+          headSymName = GHC.getName <$> findHeadSymbol originalExpr
+          isTrueOrFalse = case headSymName of
+            Nothing -> False
+            -- We have to do it this convoluted way because GHC decides that
+            -- `nm` is a *variable* but `'True` and `'False` are data constructors and
+            -- they are not equal as names even if `originalExpr` is e.g. the expression `True`
+            Just nm ->
+              or [ GHC.getOccString nm == GHC.getOccString con
+                 | Just con <- [ GHC.getName <$> Map.lookup sym nameInfo | sym <- ['True, 'False] ] -- TODO: likewise to above these two should always both return `Just`
+                 , GHC.nameSpacesRelated (GHC.occNameSpace . GHC.getOccName $ nm) (GHC.occNameSpace . GHC.getOccName $ con)
+                 ]
+      if tyHeadName /= bool || isTrueOrFalse
+      then return compiledTerm
+      -- It is a boolean that's not `True` or `False`,
+      -- generate the code:
+      -- ```
+      -- if compiledTerm
+      -- then trace (compiledTerm was true) True
+      -- else trace (compiledTerm was false) False
+      -- ```
+      else do
```

Comment:
I really don't love how much code this is, and a lot of it is duplicating the code for compiling case expressions. It's going to be annoying to maintain. I wonder whether it would be better to create the corresponding Core expression and then call `compile` on that... Maybe we could even write a _function_ and insert a call to that? e.g.

```
traceBool :: BuiltinString -> BuiltinString -> Bool -> Bool
traceBool trueLabel falseLabel c = if c then trace trueLabel True else trace falseLabel False
```

Reply:
Right, you would compile `traceBool` and apply the result to the strings and the `compiledCode` argument to `coverageCompile` as plutus code.

Reply:
Yeah, and if we're just making an application we could probably do that either by making a Core expression or by directly making a PIR application.

Reply:
We can't do it in core because that would make `coverageCompile` non-compositional.

Reply:
So PIR it is.